Parcial 1 Minería de datos

In [ ]:
import pandas as pd
import numpy as np
import csv

In [ ]:
pd.set_option('display.max_columns', None) #Mostrar todas las columnas


In [ ]:
df = pd.read_csv(
    'https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/student-mat.csv',
    sep = ";")
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         6   5   6   6  
1     1       3         4   5   5   6  
2     3       3        10   7   8  10  
3     1       5         2  15  14  15  
4     2       5         4   6  10  10

In [ ]:
df.isna().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

In [ ]:
df.isnull().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

In [ ]:
pd.unique(df['address'])

array(['U', 'R'], dtype=object)

In [ ]:
pd.isnull(df)

school    sex    age  address  famsize  Pstatus   Medu   Fedu   Mjob  \
0     False  False  False    False    False    False  False  False  False   
1     False  False  False    False    False    False  False  False  False   
2     False  False  False    False    False    False  False  False  False   
3     False  False  False    False    False    False  False  False  False   
4     False  False  False    False    False    False  False  False  False   
..      ...    ...    ...      ...      ...      ...    ...    ...    ...   
390   False  False  False    False    False    False  False  False  False   
391   False  False  False    False    False    False  False  False  False   
392   False  False  False    False    False    False  False  False  False   
393   False  False  False    False    False    False  False  False  False   
394   False  False  False    False    False    False  False  False  False   

      Fjob  reason  guardian  traveltime  studytime  failures  schoolsup  \
0    False   False     False       False      False     False      False   
1    False   False     False       False      False     False      False   
2    False   False     False       False      False     False      False   
3    False   False     False       False      False     False      False   
4    False   False     False       False      False     False      False   
..     ...     ...       ...         ...        ...       ...        ...   
390  False   False     False       False      False     False      False   
391  False   False     False       False      False     False      False   
392  False   False     False       False      False     False      False   
393  False   False     False       False      False     False      False   
394  False   False     False       False      False     False      False   

     famsup   paid  activities  nursery  higher  internet  romantic  famrel  \
0     False  False       False    False   False     False     False   False   
1     False  False       False    False   False     False     False   False   
2     False  False       False    False   False     False     False   False   
3     False  False       False    False   False     False     False   False   
4     False  False       False    False   False     False     False   False   
..      ...    ...         ...      ...     ...       ...       ...     ...   
390   False  False       False    False   False     False     False   False   
391   False  False       False    False   False     False     False   False   
392   False  False       False    False   False     False     False   False   
393   False  False       False    False   False     False     False   False   
394   False  False       False    False   False     False     False   False   

     freetime  goout   Dalc   Walc  health  absences     G1     G2     G3  
0       False  False  False  False   False     False  False  False  False  
1       False  False  False  False   False     False  False  False  False  
2       False  False  False  False   False     False  False  False  False  
3       False  False  False  False   False     False  False  False  False  
4       False  False  False  False   False     False  False  False  False  
..        ...    ...    ...    ...     ...       ...    ...    ...    ...  
390     False  False  False  False   False     False  False  False  False  
391     False  False  False  False   False     False  False  False  False  
392     False  False  False  False   False     False  False  False  False  
393     False  False  False  False   False     False  False  False  False  
394     False  False  False  False   False     False  False  False  False  

[395 rows x 33 columns]

In [ ]:
df.isna

<bound method DataFrame.isna of     school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     reason guardia

**1 punto**. En ninguna variable del dataset existen valores nulos o vacíos.

In [51]:
df['address'].value_counts()
df.groupby(['address']).count()



school  sex  age  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  reason  \
address                                                                       
R            88   88   88       88       88    88    88    88    88      88   
U           307  307  307      307      307   307   307   307   307     307   

         guardian  traveltime  studytime  failures  schoolsup  famsup  paid  \
address                                                                       
R              88          88         88        88         88      88    88   
U             307         307        307       307        307     307   307   

         activities  nursery  higher  internet  romantic  famrel  freetime  \
address                                                                      
R                88       88      88        88        88      88        88   
U               307      307     307       307       307     307       307   

         goout  Dalc  Walc  health  absences   G1   G2   G3  
address                                                      
R           88    88    88      88        88   88   88   88  
U          307   307   307     307       307  307  307  307

In [49]:
hombress = df[df['sex'] == 'M']
hombress.groupby(['address']).count()

school  sex  age  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  reason  \
address                                                                       
R            44   44   44       44       44    44    44    44    44      44   
U           143  143  143      143      143   143   143   143   143     143   

         guardian  traveltime  studytime  failures  schoolsup  famsup  paid  \
address                                                                       
R              44          44         44        44         44      44    44   
U             143         143        143       143        143     143   143   

         activities  nursery  higher  internet  romantic  famrel  freetime  \
address                                                                      
R                44       44      44        44        44      44        44   
U               143      143     143       143       143     143       143   

         goout  Dalc  Walc  health  absences   G1   G2   G3  
address                                                      
R           44    44    44      44        44   44   44   44  
U          143   143   143     143       143  143  143  143

**Punto 2** Hay 307 estudiantes que viven en el área urbana y 88 que viven en el área rural los cuales 44 son hombres.

In [61]:
fallas = df[df['absences'] > 5]
fallas.groupby(['address']).count()


school  sex  age  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  reason  \
address                                                                       
R            38   38   38       38       38    38    38    38    38      38   
U           108  108  108      108      108   108   108   108   108     108   

         guardian  traveltime  studytime  failures  schoolsup  famsup  paid  \
address                                                                       
R              38          38         38        38         38      38    38   
U             108         108        108       108        108     108   108   

         activities  nursery  higher  internet  romantic  famrel  freetime  \
address                                                                      
R                38       38      38        38        38      38        38   
U               108      108     108       108       108     108       108   

         goout  Dalc  Walc  health  absences   G1   G2   G3  
address                                                      
R           38    38    38      38        38   38   38   38  
U          108   108   108     108       108  108  108  108

In [63]:
fallasH = fallas[fallas['sex'] == 'M']
fallasH.groupby(['address']).count()


school  sex  age  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  reason  \
address                                                                       
R            18   18   18       18       18    18    18    18    18      18   
U            48   48   48       48       48    48    48    48    48      48   

         guardian  traveltime  studytime  failures  schoolsup  famsup  paid  \
address                                                                       
R              18          18         18        18         18      18    18   
U              48          48         48        48         48      48    48   

         activities  nursery  higher  internet  romantic  famrel  freetime  \
address                                                                      
R                18       18      18        18        18      18        18   
U                48       48      48        48        48      48        48   

         goout  Dalc  Walc  health  absences  G1  G2  G3  
address                                                   
R           18    18    18      18        18  18  18  18  
U           48    48    48      48        48  48  48  48

**3 punto** Hay 38 estudiantes que viven en la zona rural y tienen más de 5 fallas, los cuales 18 son hombres.

In [68]:
GP = df[df['school'] == 'GP']
GP.groupby(['internet']).mean() 

age      Medu      Fedu  traveltime  studytime  failures  \
internet                                                                   
no        16.673077  2.365385  2.269231    1.576923   1.923077  0.346154   
yes       16.494949  2.878788  2.602694    1.353535   2.087542  0.313131   

            famrel  freetime     goout      Dalc      Walc    health  \
internet                                                               
no        3.942308  3.115385  2.923077  1.326923  2.173077  3.750000   
yes       3.962963  3.242424  3.144781  1.464646  2.276094  3.545455   

          absences         G1         G2         G3  
internet                                             
no        3.980769  10.538462   9.903846   9.673077  
yes       6.313131  11.010101  10.936027  10.632997

In [69]:
MS = df[df['school'] == 'MS']
MS.groupby(['internet']).mean() 

age      Medu     Fedu  traveltime  studytime  failures  \
internet                                                                  
no        18.285714  1.857143  2.00000    1.785714   1.928571  0.785714   
yes       17.906250  2.562500  2.40625    1.968750   1.781250  0.312500   

            famrel  freetime     goout      Dalc      Walc  health  absences  \
internet                                                                       
no        3.642857  3.142857  2.928571  1.714286  2.571429   4.000  3.571429   
yes       3.906250  3.406250  3.156250  1.781250  2.500000   3.125  3.843750   

                 G1        G2         G3  
internet                                  
no         9.785714   9.00000   8.428571  
yes       11.062500  10.71875  10.468750

**4 PUNTO** Para el colegio Gabriel Pereira la media de las notas del tercer corte de los estudiantes que si tienen internet es de 10.63 y de los que no 9.67, por otro lado, para el colegio usinho de Silveira la media de las notas del tercer cortes de de los estudiantes que si tienen internet es de 10.46 y de los que no 8.42.

In [28]:
G1 = list(df['G1'])
G2 = list(df['G2'])
G3 = list(df['G3'])
nota_final=list()
for i in range (len(G1)):
    valor = (G1[i]*0.30)+(G2[i]*0.30)+(G3[i]*0.40)
    nota_final.append(valor)
df1 =pd.DataFrame( {'Nota 1 ' : G1,
          'Nota 2' : G2, 'Nota 3':G3, 'Nota final' : nota_final})
df1.describe()

Nota 1       Nota 2      Nota 3  Nota final
count  395.000000  395.000000  395.000000  395.000000
mean    10.908861   10.713924   10.415190   10.652911
std      3.319195    3.761505    4.581443    3.769058
min      3.000000    0.000000    0.000000    1.200000
25%      8.000000    9.000000    8.000000    8.300000
50%     11.000000   11.000000   11.000000   10.700000
75%     13.000000   13.000000   14.000000   13.300000
max     19.000000   19.000000   20.000000   19.400000

In [29]:
df1.median()

Nota 1        11.0
Nota 2        11.0
Nota 3        11.0
Nota final    10.7
dtype: float64

**5 punto**. Como se evidencia en la tabla anterior, la media de la nota final de los 395 estudiantes es de 10.652911, la máxima nota alcanzada es de 19.4, la minima de 1.2 y la mediana de 10.7.

**6 punto.** En mi opinión, realizaría un tipo de análisis de clasificación puesto que La clasificación nos ayuda a entrenar al algoritmo para clasificar los datos de entrada en variables discretas, donde mi variable discreta seria si los estudiantes aprobaron el bachiderato o no.